### PR TITLE
Assignment#2: Bank ATM's Management service

### PR DESCRIPTION
**Pull request by istocoana**
### Summary 
**Bank Account**

- individual bank account which has attributes like name, job, email, and address.
- the initialize method sets the initial values for these attributes when a new account is created (constructor). The initial balance of a new user is set to 0. 
- each new user can deposit money in their account, increasing the balance with unlimited money.
- users can withdraw money from their account while enforcing a maximum daily withdrawal limit of $5000. It provides feedback to the user for three scenarios: 

1. If the withdrawal amount exceeds the maximum daily limit, it displays an error message regarding the limit.
2. If the withdrawal amount is greater than the available account balance, it shows an error message indicating insufficient funds.
3. If the withdrawal is valid and there are sufficient funds, it deducts the specified amount from the account balance and informs the user about the successful transaction and the updated account balance.

**Bank System**

- creates an instance variable representing a list of bank accounts. It is initialized as an empty array when a new BankSystem object is created.
- the create_account method allows the creation of a new bank account within the system. It accepts four parameters name, function, email and address creating a new BankAccount object with the details provided and adding it to the list of accounts (@accounts). When the account is created a message is displayed that the account has been successfully created returning the account.

**Main**
- this is a menu of the program.
- 3 initial users are created to show each case of the program when users withdraw money (user1, user2, user3).  
- user1 is the one whose case is acceptable without any error message, user2 tries to withdraw money but he doesn't have enough, and user3's withdrawal exceeds the $5000 per day limit.
- Next, the program prompts the user to create their bank account by entering their name, job, email, and address. The provided details are then used to create a new account using the create_account method.
- The program enters a loop where the user can interact with their account. The user can choose to deposit money, withdraw money, check their balance, or exit the program.
- If the user chooses to deposit or withdraw money, they are prompted to enter the amount they want to deposit or withdraw. The corresponding methods (deposit_money and withdraw_money) are called on the user's account to handle the transactions.
- If the user chooses to check their balance, the program displays their current account balance.
- If the user chooses to exit, the program displays a goodbye message and terminates the loop, effectively ending the program.

**Unit Tests**
- I created a file for each class, bank_account, and bank_system.

_bank_account_spec.rb:_
- let(:bank_account) defines a bank_account variable that is initialized with a new BankAccount object having default attributes.
- describe '#initialize' contains a single test that checks if the initial balance of a newly created BankAccount object is set to 0.
- describe '#deposit_money' contains a test that checks if the deposit_money method adds money to the account balance correctly.
- describe '#withdraw_money': contains tests for different withdrawal scenarios:
1. A test to ensure that a valid withdrawal reduces the account balance correctly.
2. A test to check if attempting to withdraw an amount exceeding the maximum daily limit results in the appropriate error message.
3. A test to check if attempting to withdraw more than the account balance results in the appropriate error message.

_bank_system_spec.rb:_
- describe '#initialize' contains a single test that checks if a new instance of BankSystem initializes with an empty array for the accounts attribute.
-describe '#create_account' test that checks if the create_account method successfully adds a new account to the accounts list of the BankSystem object. The test creates a new BankSystem object, then calls the create_account method with sample details and verifies that the created account is included in the accounts list.

### CODE REVIEW

File: lib/bank_account.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      puts "The maximum daily withdrawal amount has been exceeded. You can withdraw a maximum of $5000."
+    elsif money > @balance
+      puts "Your balance is #{@balance}. You don't have enough funds in your account to withdraw $#{money}" 
+    
+    else 
+      @balance -= money
+      puts "Retrieved $ #{money}. Your account balance is $ #{@balance}."
+    end 
+  end
+end
```
Comment (bogdan-boer) [inline lib/bank_account.rb]: You are missing an empty line at the end of file

File: lib/bank_account.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  def initialize(name, job, email, address)
+    @name = name
+    @job = job
+    @email = email 
+    @address = address 
+    @balance = 0
+  end
+
+  def deposit_money(money)
```
Comment (bogdan-boer) [inline lib/bank_account.rb]: What happens if the user inputs an negative amount?

File: lib/bank_account.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    @balance += money 
+    puts "Deposited $ #{money}. Your account balance is $ #{@balance}."
+  end
+
+  def withdraw_money(money)
```
Comment (bogdan-boer) [inline lib/bank_account.rb]: Same as above

File: lib/bank_account.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+class BankAccount
+ 
+  attr_accessor :name, :job, :email, :address, :balance
```
Comment (bogdan-boer) [inline lib/bank_account.rb]: `:name, :job, :email, :address` sound more like fields for an User than a bank account, you should create a new model for these attributes

File: spec/bank_account_spec.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
+require_relative '../lib/bank_account'
+
+RSpec.describe BankAccount do
+
+  let(:bank_account) {
+     BankAccount.new("Name", "Job", "Email", "Address") 
+
```
Comment (bogdan-boer) [inline spec/bank_account_spec.rb:9]: You have an extra line here

File: lib/bank_account.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @balance = 0
+  end
+
+  def deposit_money(money)
+    @balance += money 
+    puts "Deposited $ #{money}. Your account balance is $ #{@balance}."
+  end
+
+  def withdraw_money(money)
+    if money > 5000
```
Comment (VictorMotogna) [inline lib/bank_account.rb]: The assignment was to check that in 1 day all withdrawals cannot pass 5000$. If I understand correctly, using your approach, I will be able to extract 4.999$ multiple times in the same day.

Comment (VictorMotogna) [reply lib/bank_account.rb]: I think it would be a good idea to also add a test for this use case

File: lib/user.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  def initialize(name, job, email, address, pin)
+    @name = name
+    @job = job
+    @email = email
+    @address = address
+    set_pin(pin)
+    @logged_in = false
+  end
+
+  def set_pin(pin)
```
Comment (butiricristian) [inline lib/user.rb]: Could this be changed to `def pin=`?

File: lib/user.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    if valid_pin?(pin)
+      @pin = pin
+      puts "PIN set successfully. [#{pin}]"
+      true
+    else
+      puts "Invalid PIN. It must be a 4-digit number."
+      false
+    end
```
Comment (butiricristian) [inline lib/user.rb]: I would suggest checking for errors first and returning from the method when finding one. This way we can avoid if-else clauses:
```suggestion
    unless valid_pin?(pin)
      puts "Invalid PIN. It must be a 4-digit number."
      return false
    end
    
    @pin = pin
    puts "PIN set successfully. [#{pin}]"
    true
```

File: lib/main.rb [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+  user_accounts = {}
+
+  default_user1 = User.new("John Doe", "Engineer", "john.doe@example.com", "123 Elm Street, Cityville, USA", "1111")
+  user_account1 = bank_system.create_account(default_user1)
+  user_account1.balance = 1000
+  puts "Welcome, #{default_user1.name}! This is your balance: #{user_account1.balance}."
+  user_account1.deposit_money(3000)
+  user_account1.withdraw_money(2000)
+  user_accounts[user_account1.user.name] = user_account1
```
Comment (butiricristian) [inline lib/main.rb]: The main method seems to be getting pretty long. Could we maybe move the user creation into their separate functions: `def create_user1`, `def create_user2`, etc.?

File: modules/transaction.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+module Transaction
+  def deposit_money(money)
+      if money > 0
+        @balance += money 
+        puts "Deposited $ #{money}. Your account balance is $ #{@balance}."
+      else  
+        puts "You can't deposit a negative amount!"
+      end
```
Comment (butiricristian) [inline modules/transaction.rb]: Here we can also do the negative check first and return early from the method:
```suggestion
      if money <= 0
        puts "You can't deposit a negative amount!"
        return
      end
```

File: spec/bank_system_spec.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  
+  describe '#initialize' do
+    it 'creates an empty array of accounts' do
+      expect(bank_system.accounts).to eq([])
+    end
+  end
+
+  describe '#create_account' do
+    it 'adds a new account to the accounts list' do
+      expect(BankAccount).to receive(:new).with(User).and_return(User)
```
Comment (butiricristian) [inline spec/bank_system_spec.rb]: I think you need to replace User with a mocked instance here. User is the class name and BankAccount's initialize should receive a User instance not the User class.

File: spec/create_user_spec.rb [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  it 'retries if an invalid PIN is provided at first' do
+    invalid_pin = 'invalid'
+    valid_pin = user_input_values[:pin]
+
+    allow_any_instance_of(Kernel).to receive(:gets).and_return(invalid_pin, valid_pin)
+
+    new_user = subject.create_user
+    expect(new_user).to be_a(User)
+    expect(new_user.pin).to eq(user_input_values[:pin])
+  end
```
Comment (butiricristian) [inline spec/create_user_spec.rb]:
```suggestion
context 'when an invalid PIN is provided' do
  let(:invalid_pin) { 'invalid' }

  ...
end
```

File: spec/transaction_spec.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    attr_accessor :balance
+
+    def initialize(initial_balance)
+      @balance = initial_balance
+    end
+  end
+
+  describe '#deposit_money' do
+    it 'adds money to the balance' do
+      test_instance = TransactionTest.new(1000)
```
Comment (butiricristian) [inline spec/transaction_spec.rb:19]: It's a good practice to define variables with `let` when writing tests and to do any manipulation in `before` methods. This way we separate definitions and actions from expectations

**Review by butiricristian — CHANGES_REQUESTED** [review]
Left a couple of remarks, but looks good. Nice job 👍

File: lib/main.rb [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    choice = gets.chomp.to_i 
+
+    case choice 
+    when 1
+      include CreateUser
+      user_account = bank_system.create_account(create_user)
+      user_accounts[user_account.user.name] = user_account
+    when 2
+      puts "Please enter your name:"
+      name = gets.chomp
```
Comment (florinionce) [inline lib/main.rb:48]: the logic from each action should be extracted

File: spec/bank_system_spec.rb [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    it 'creates an empty array of accounts' do
+      expect(bank_system.accounts).to eq([])
+    end
+  end
+
+  describe '#create_account' do
+    it 'adds a new account to the accounts list' do
+      user_instance = instance_double(User, name: 'John Doe')
+      expect(BankAccount).to receive(:new).with(user_instance).and_return(user_instance)
+      bank_system.create_account(user_instance)
```
Comment (florinionce) [inline spec/bank_system_spec.rb:18]: you could use `subjects`
https://www.betterspecs.org/#subject

**Pull request merged**